src/components/QueryBuilder/index.js

```diff
@@ -0,0 +1,298 @@
+            onClick={() => handleQueryDelete(selectedQuery.id)}
+          >
+            <MdDelete height={18} width={18} />
+          </button>
+        )}
+        <QueryBuilderActionMenu
+          options={options}
+          onOptionClick={handleQueryBuilderAction}
+        />
+      </div>
+      <Modal
+        open={modalIsOpen}
+        onClose={() => setModalIsOpen(false)}
+        title="Save Current Query"
+        size="small"
+      >
+        <div className="flex flex-col pt-5 pb-3 max-w-lg" style={{}}>
+          <input
+            type="text"
+            defaultValue={savedQueryValue}
+            className="w-full shadow-sm focus:ring-indigo-500 focus:border-indigo-400 block py-2 sm:text-sm border-gray-300 rounded-md mr-2 mb-2"
+            style={{ minWidth: "300px" }}
+            placeholder="Query name"
+            onChange={(e) => setSavedQueryValue(e.target.value)}
+          />
+          <div className="flex justify-end mt-3">
+            <button
+              className={`border border-gray-300 rounded-md px-3 py-2 text-sm whitespace-nowrap bg-indigo-700 text-gray-50 cursor-pointer}`}
+              type="button"
+              onClick={() => saveQuery()}
+            >
+              Save
+            </button>
+          </div>
+        </div>
+      </Modal>
+    </div>
+  );
+};
+
+function QueryBuilderActionMenu({ onOptionClick, options }) {
+  return (
+    <Menu as="div" className="relative inline-block text-left">
+      <div>
+        <Menu.Button className="inline-flex justify-center w-full p-3 bg-white text-sm font-medium text-gray-700">
+          <BiCog className="content-center" />
+        </Menu.Button>
+      </div>
+      <Transition
+        as={Fragment}
+        enter="transition ease-out duration-100"
+        enterFrom="transform opacity-0 scale-95"
+        enterTo="transform opacity-100 scale-100"
+        leave="transition ease-in duration-75"
+        leaveFrom="transform opacity-100 scale-100"
+        leaveTo="transform opacity-0 scale-95"
+      >
+        <Menu.Items className="origin-top-right absolute right-0 mt-2 w-56 rounded-md shadow-lg bg-white ring-1 ring-black ring-opacity-5 focus:outline-none z-10">
+          <div className="py-1">
+            {options.map((option) => {
+              return (
+                <Menu.Item key={option.id}>
+                  {({ active }) => (
+                    <div
+                      className={clsx(
+                        active ? "bg-gray-100 text-gray-900" : "text-gray-700",
+                        "block px-4 py-2 text-sm",
+                        "cursor-pointer"
+                      )}
+                      onClick={() => onOptionClick(option)}
+                    >
+                      {option.label}
+                    </div>
+                  )}
+                </Menu.Item>
+              );
+            })}
+          </div>
+        </Menu.Items>
+      </Transition>
+    </Menu>
+  );
+}
```

Comment:
Does it make sense to have separate component for query actions? Haven't tried coding so not sure how it would look.